### PR TITLE
Add reverse translation.

### PR DESCRIPTION
**Pull request by DailyDreaming**
Tests have been added.

### CODE REVIEW

**Comment by DailyDreaming** [comment]
@xbrianh This should be ready for review.

**Comment by DailyDreaming** [comment]
@xbrianh This should be ready for review.

File: herzog/__init__.py [hunk 1]
```diff
@@ -38,10 +39,51 @@ def __exit__(self, *args, **kwargs):
...
     with open(os.path.join(os.path.dirname(__file__), "data", "python_3_boiler.json")) as fh:
         boiler = json.loads(fh.read())
-    ipynb = dict(cells=cells, **boiler)
+    ipynb = json.dumps(dict(cells=cells, **boiler), indent=indent)
     return ipynb
+
+def translate_to_herzog(ipynb_handle: TextIO, indent: int = 4) -> str:
+    cells = load_ipynb_cells(ipynb_handle)
+    python_prefix = ' ' * indent
+    markdown_prefix = python_prefix + '# '
```
Comment (xbrianh) [inline herzog/__init__.py:73]: Herzog uses docstrings to encode markdown, not comments.

See the [readme example](https://github.com/xbrianh/herzog#example)

File: tests/test_herzog.py [hunk 2]
```diff
@@ -26,7 +37,62 @@ def test_parser(self):
...
         for expected_cell, cell in zip(expected_cells, cells):
-            self.assertEqual(expected_cell, cell)
+
+            # convert all "source" to "string"; equally valid as a list or a string, example:
+            #     '# This is a header\ndoom and gloom\n\n## frank is a gangster\nevidence'
+            #     ['# This is a header\n', 'doom and gloom\n', '\n', '## frank is a gangster\n', 'evidence']
+            if isinstance(expected_cell.get('source', None), list):
+                expected_cell['source'] = ''.join(expected_cell['source'])
+
+            self.assertEqual(expected_cell, cell, f'\n{expected_cell}\n{cell}')
```
Comment (xbrianh) [inline tests/test_herzog.py:47]: I’m not sure I understand why this test was modified.

File: herzog/__init__.py [hunk 3]
```diff
@@ -38,10 +39,51 @@ def __exit__(self, *args, **kwargs):
...
+def translate_to_ipynb(herzog_handle: TextIO, indent: int = 2) -> str:
+    cells = [obj.to_ipynb_cell() for obj in parse_cells(herzog_handle)
              if obj.has_ipynb_representation]
     with open(os.path.join(os.path.dirname(__file__), "data", "python_3_boiler.json")) as fh:
         boiler = json.loads(fh.read())
-    ipynb = dict(cells=cells, **boiler)
+    ipynb = json.dumps(dict(cells=cells, **boiler), indent=indent)
     return ipynb
+
+def translate_to_herzog(ipynb_handle: TextIO, indent: int = 4) -> str:
```
Comment (xbrianh) [inline herzog/__init__.py:70]: It’s temping to have this method return a list of lines, or make it a generator over lines. This would make the Python API more friendly IMO, and matches the programmatic style of the herzog/parser.py module.

File: herzog/parser.py [hunk 4]
```diff
@@ -26,8 +26,10 @@ def __init__(self, cell_type: CellType, lines: Iterable[str]):
                 else:
                     self.lines.append(line)
             elif CellType.markdown == self.cell_type:
-                if '"""' == line:
+                if '"""' == line or "pass" == line:
                     pass
+                elif line.startswith('#'):
+                    self.lines.append(line[1:].strip())
```
Comment (xbrianh) [inline herzog/parser.py:32]: This change to Herzog syntax should be moved into a separate PR. It also should be motivated by explaining the purpose of the change.

File: herzog/parser.py [hunk 5]
```diff
@@ -92,12 +94,14 @@ def _validate_cell(cell_lines: List[str], line_number: Optional[int]=None):
...
+def parse_cell_type(s: str) -> str:
+    return s.strip()[len("with herzog.Cell("):-len('):')].strip().strip('"').strip("'").strip()
+
 def parse_cells(raw_lines: TextIO) -> Generator[HerzogCell, None, None]:
     rlines = _RewindableIterator(raw_lines)
     for line in rlines:
         if "with herzog.Cell" in line:
-            cell_type_string = line.strip()[len("with herzog.Cell"):].strip('"():')
-            cell_type = CellType[cell_type_string]
+            cell_type = CellType[parse_cell_type(line)]
```
Comment (xbrianh) [inline herzog/parser.py:104]: What’s the purpose of the convenience method? Was there a bug, or is this a purely cosmetic change?

In either case this should be split out into a separate PR.

**Review by xbrianh — CHANGES_REQUESTED** [review]
Thanks for updating the CLI interface! I few more things to clean up:
  - There appears to be changes to Herzog syntax for markdown. The motivation for this should be explained, and it should be split out into a separate PR.
  - The inverse parser appears incomplete since iPython magic is not handled (e.g. `%pip install foo`). The forward parser [handles these](https://github.com/xbrianh/herzog/blob/d9444e7dcad4a313456f9236efd1cda9f8bd65fc/herzog/parser.py#L24)
  -  The inverse parser should add a blank line between context managers. When I tried it out there was no whitespace between cells.

File: herzog/__init__.py [hunk 6]
```diff
@@ -38,10 +39,51 @@ def __exit__(self, *args, **kwargs):
                     del __main__.__dict__[key]
             __main__.__dict__.update(self._state_modules)
 
-def generate(handle: TextIO):
-    cells = [obj.to_ipynb_cell() for obj in parse_cells(handle)
+def load_ipynb_cells(ipynb: TextIO) -> List[Dict[Any, Any]]:
+    try:
+        cells = json.loads(ipynb.read()).get('cells', None)
+    except Exception:
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply herzog/__init__.py:45]: I figured any error opening the file and reading it would indicate that it wasn't a ipynb file, though I guess it's a little broad.  Will change to `JSONDecodeError`.

File: tests/fixtures/example.py [hunk 7]
```diff
@@ -8,13 +8,12 @@
...
-    doom and gloom
+    # # This is a header
+    # doom and gloom
 
-    ## frank is a ganster
-    evidence
-    """
+    # ## frank is a gangster
+    # evidence
+    pass
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply tests/fixtures/example.py:16]: `#` is more deterministic than `"""`.  Under the old method, anything that literally contained `""""` in the markdown file would not be treated correctly.  For example, the following perfectly valid markdown:

```
Very
"""
Cool
"""
Title.
```

Doesn't work with the current method.  It also doesn't handle `'''`, so I switched the method to something that easily comes on and off, deterministically, without specific parsing.

I'll include all of this in a separate PR, as per an earlier comment.

File: herzog/__init__.py (hunk 1 above)
Comment (DailyDreaming) [reply herzog/__init__.py:73]: I'm concerned that the current docstring implementation may not be deterministic.  I'll make a separate PR where I hope to make my points clear.

File: tests/test_herzog.py (hunk 2 above)
Comment (DailyDreaming) [reply tests/test_herzog.py:47]: I'm not sure if it's the ipynb version, but I believe I explained it in the comments.

File: herzog/__init__.py (hunk 3 above)
Comment (DailyDreaming) [reply herzog/__init__.py:70]: This was an effort to match current behavior.  I'll make an attempt.  I would also support a follow-up PR, if you wanted to make one, to bring code into how you see it should be styled.

File: herzog/parser.py (hunk 5 above)
Comment (DailyDreaming) [reply herzog/parser.py:102]: There was a bug.  Herzog breaks if you use single quotes.

Comment (DailyDreaming) [reply herzog/parser.py:102]: Now fixed in #22.

File: herzog/parser.py (hunk 4 above)
Comment (DailyDreaming) [reply herzog/parser.py:30]: Opened in #27.

File: herzog/__init__.py (hunk 1 above)
Comment (DailyDreaming) [reply herzog/__init__.py:73]: Opened in #27.

File: tests/fixtures/example.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply tests/fixtures/example.py:16]: Opened in #27.

**Comment by DailyDreaming** [comment]
@xbrianh All comments should be addressed.  There was a question to explain why a test was changed and the following hopefully adequately explains it:

```
                    # convert all "source" to "string"; equally valid as a list or a string, example:
                    #     '# This is a header\ndoom and gloom\n\n## frank is a gangster\nevidence'
                    #     ['# This is a header\n', 'doom and gloom\n', '\n', '## frank is a gangster\n', 'evidence']
```

If you meant something else let me know.

**Comment by DailyDreaming** [comment]
@xbrianh I missed something.  Will wait to hear back before committing in case you're reviewing.

File: tests/test_herzog.py (hunk 2 above)
Comment (xbrianh) [reply tests/test_herzog.py:40]: Is the comment meant to indicate that .ipynb formatted cells can contain either lists or strings in the `source` field? Can you provide a link to documentation describing this feature of the .ipynb format?

Currently, Herzog does not produce cells containing list formatted `source` fields. Until it does so, it seems more appropriate for the test to expect only string formatted `source` fields.

Unless I’m missing something?

**Comment by xbrianh** [comment]
@DailyDreaming 

Has this been handled?:
> The inverse parser appears incomplete since iPython magic is not handled (e.g. %pip install foo). The forward parser handles these

**Comment by DailyDreaming** [comment]
@xbrianh Should be addressed.

File: herzog/parser.py [hunk 8]
```diff
@@ -19,23 +19,23 @@ def __init__(self, cell_type: CellType, lines: Iterable[str]):
         self.lines: List[str] = list()
         for line in lines:
             if CellType.python == self.cell_type:
-                if "pass" == line:
+                if 'pass' == line:
```
Comment (xbrianh) [inline herzog/parser.py:22]: Please revert this line, and use double quotes throughout to match style with the rest of herzog:
  - double quotes for strings
  - single quotes for keys

File: herzog/parser.py [hunk 9]
```diff
@@ -67,7 +67,7 @@ def __iter__(self):
         except StopIteration:
             pass
 
-    def rewind(self):
+    def rewind(self) -> None:
```
Comment (xbrianh) [inline herzog/parser.py:70]: Please omit type hints for functions or methods returning `None`. This should be applied through this PR.

File: tests/test_herzog.py (hunk 2 above)
Comment (xbrianh) [reply tests/test_herzog.py:40]: Please break this out in a separate PR.

**Comment by DailyDreaming** [comment]
@xbrianh Let me know if/when you're good with me pushing one more commit.

File: herzog/__init__.py [hunk 10]
```diff
@@ -38,10 +39,39 @@ def __exit__(self, *args, **kwargs):
...
+
+    for cell in cells:
+        if isinstance(cell.get('source', None), list):
+            cell['source'] = ''.join(cell['source'])
+
+        if cell['cell_type'] == 'markdown':
+            yield "\nwith herzog.Cell('markdown'):"
+            yield '\n    """\n'
+            if cell['source'].startswith('%') or cell['source'].startswith('!'):
+                yield textwrap.indent('#' + cell['source'], prefix=prefix).rstrip()
```
Comment (xbrianh) [inline herzog/__init__.py:69]: Please use the constants Herzog [defines](https://github.com/xbrianh/herzog/blob/5c16506487943bc7f0b8c642bfa143c3ed716a85/herzog/parser.py#L11-L12) for magic prefixes.

e.g.
```
if cell[‘source’].startswith(“%”):
    source = cell[‘source’].replace(“#”, parser.JUPYTER_MAGIC_PFX, 1)
```

File: herzog/__init__.py [hunk 11]
```diff
@@ -38,10 +39,39 @@ def __exit__(self, *args, **kwargs):
...
+            else:
+                yield textwrap.indent(cell['source'], prefix=prefix).rstrip()
+            yield '\n    """\n'
+        elif cell['cell_type'] == 'code':
+            yield "\nwith herzog.Cell('python'):\n"
+            yield textwrap.indent(cell['source'], prefix=prefix).rstrip()
+        else:
+            raise NotImplementedError(f"cell_type not implemented yet: {cell['cell_type']}")
```
Comment (xbrianh) [inline herzog/__init__.py:77]: What’s the correct answer for this case?:
  - raising
  - warning and continuing (e.g., doing the best we can)

Since reverse translation will likely be used to convert existing notebooks into the Herzog format, the second case may be the correct answer. What do you think?

**Review by xbrianh — CHANGES_REQUESTED** [review]
This is looking very close! Just a few concerns about style and maintainability, please see comments.

File: herzog/__init__.py (hunk 11 above)
Comment (DailyDreaming) [reply herzog/__init__.py:77]: @xbrianh I think that other cases may have unexpected behavior and a clear error here will make debugging easier if it's, say, in `R`.

Comment (xbrianh) [reply herzog/__init__.py:81]: - A warning can also be clear
- An error prevents output. With warnings, users still get _some_ output

Comment (DailyDreaming) [reply herzog/__init__.py:81]: @xbrianh Alright, will change it to a warning.

**Comment by DailyDreaming** [comment]
@xbrianh All comments should be addressed.

**Comment by DailyDreaming** [comment]
Forgot one ` -> None`.  Ironically the one you commented on.  Let me know and I'll push the commit. 

Otherwise, review comments should all be addressed.

**Comment by DailyDreaming** [comment]
It's been a few hours so I assume you're not currently in the middle of reviewing and it's okay to add the commit.  Am going to watch a movie and if I don't hear back by then, I'm assuming it's cool to push.

File: scripts/herzog [hunk 12]
```diff
@@ -1,14 +1,60 @@
...
+    if len(args) == 1 and args[0] not in ("--help", "-h", "--version", "-v"):
+        # maintain backwards compatibility
+        with open(args[0]) as fh:
+            print(json.dumps(herzog.translate_to_ipynb(fh), indent=2))
+            exit()
+
+    parser = argparse.ArgumentParser(prog="Herzog",
+                                     description="Developer tool to translate notebooks to runnable code and "
+                                                 "back again, allowing notebooks to be written, reviewed, "
+                                                 "and versioned-controlled as normal code.")
```
Comment (xbrianh) [inline scripts/herzog:28]: While I appreciate the usage recommendations, I prefer a more purely functional description of Herzog.

Perhaps something like this?:
> Convert Herzog formatted Python scripting to iPython .ipynb files, and back again.

File: herzog/__init__.py [hunk 13]
```diff
@@ -38,10 +39,49 @@ def __exit__(self, *args, **kwargs):
...
+                else:
+                    yield line + "\n"
+        else:
+            print(f"cell_type not implemented yet: {cell['cell_type']}", file=sys.stderr)
+            # warn the user and wrap in a quote as we would with markdown
+            s = f'\nwith herzog.Cell("{cell["cell_type"]}"):\n    """\n'
+            s += textwrap.indent(cell['source'], prefix=prefix).rstrip()
+            s += '\n    """\n'
+            for line in s.split("\n"):
+                yield line + "\n"
```
Comment (xbrianh) [inline herzog/__init__.py:87]: out-of-band discussion would be useful before introducing new features into a PR, just so we can get on the same page before implementation.

Nevertheless, this is an interesting approach. After brief consideration, I think it would be best produce commented output as shown below. Commented lines are the safest way to deal with content not understood by Herzog.
```
## .ipynb -> Herzog translation failed:
## Cell type “my-weird-cell” not supported by Herzog. Supported cell types are X, Y, Z
# with herzog.Cell(“my-weird-cell”):
#.    {cell contents}
```

**Review by xbrianh — CHANGES_REQUESTED** [review]

File: herzog/parser.py [hunk 14]
```diff
@@ -26,16 +26,16 @@ def __init__(self, cell_type: CellType, lines: Iterable[str]):
                 else:
                     self.lines.append(line)
             elif CellType.markdown == self.cell_type:
-                if '"""' == line:
+                if line in ('"""', "pass"):
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply herzog/parser.py:29]: Neither `"""` nor `pass` are a part of the markdown proper.  This is to keep `pass` from being duplicated each time (since it must be added, as it's not a part of the ipython notebook to begin with, especially one downloaded from Terra).

I'm not sure that the change makes sense outside of the context of this PR, but erring on the side of caution, I will make a new PR.

File: herzog/__init__.py [hunk 15]
```diff
@@ -38,10 +39,49 @@ def __exit__(self, *args, **kwargs):
...
+def translate_to_ipynb(herzog_handle: TextIO) -> Dict[str, Any]:
+    cells = [obj.to_ipynb_cell() for obj in parse_cells(herzog_handle)
              if obj.has_ipynb_representation]
     with open(os.path.join(os.path.dirname(__file__), "data", "python_3_boiler.json")) as fh:
         boiler = json.loads(fh.read())
-    ipynb = dict(cells=cells, **boiler)
-    return ipynb
+    return dict(cells=cells, **boiler)
+
+def translate_to_herzog(ipynb_handle: TextIO, indent: int = 4) -> Iterable[str]:
```
[the remark this answers is not in the document]
Comment (DailyDreaming) [reply herzog/__init__.py:56]: Will change.  Just want to point out that It's odd to me that functions with `yield` still give `StopIteration` errors:

```
def i():
    yield 1
    yield 2

j = i()
print(next(j))
print(next(j))
print(next(j))
```

```
1
2
Traceback (most recent call last):
  File "/home/quokka/git/herzog/i.py", line 10, in <module>
    print(next(j))
StopIteration
```

**Comment by DailyDreaming** [comment]
@xbrianh Latest comments should be addressed.  Thanks for the review.

File: herzog/__init__.py [hunk 16]
```diff
@@ -38,10 +38,53 @@ def __exit__(self, *args, **kwargs):
...
+            for line in s.split("\n"):
+                yield line + "\n"
+        elif cell['cell_type'] == "code":
+            s = "\nwith herzog.Cell('python'):\n"
+            s += textwrap.indent(cell['source'], prefix=prefix).rstrip()
+            for line in s.split("\n"):
+                if line.startswith("%"):
+                    yield line.replace("%", JUPYTER_MAGIC_PFX, 1) + "\n"
+                elif line.startswith("#"):
+                    yield line.replace("#", JUPYTER_SHELL_PFX, 1) + "\n"
```
Comment (xbrianh) [inline herzog/__init__.py:77]: The Jupyter shell prefix character is !

e.g.
```
!gsutil cp foo gs://bar
```

**Review by xbrianh — APPROVED** [review]
This LGTM, but please address the remaining comment concerning incorrect reverse translation.

File: herzog/__init__.py (hunk 16 above)
Comment (DailyDreaming) [reply herzog/__init__.py:77]: Oof, my bad.  Thanks for the catch!
